File: test/status.test.mjs
// status.test.mjs — `landfall status`, feature 20260812-010632 (US5/T048, T051).
//
// A genuine `startHookSocket`+`sendToSocket` round trip does work standalone
// in this environment (confirmed directly before writing these) — so the
// success path below binds a real socket. The fallback/absent-session paths
// deliberately do NOT need one: "no socket at all" needs only `fs.readdir`
// (always real, no I/O to a socket), and "a socket exists but nothing
// answers" is reproduced with a plain file at the expected socket path,
// which fails `net.connect` fast without anything needing to be listening —
// smaller and more deterministic than racing a real server's shutdown timing.
import test from 'node:test';
import assert from 'node:assert/strict';
import { promises as fs } from 'node:fs';
import os from 'node:os';
import path from 'node:path';

import { formatStatusLine, queryStatus, readCache, runStatus, writeCache } from '../src/status.mjs';
import { socketLocation, startHookSocket } from '../src/hooks/socket.mjs';

async function withWorkspace(fn) {
  const root = await fs.mkdtemp(path.join(os.tmpdir(), 'landfall-status-'));
  try {
    return await fn({ cwd: root, env: { ...process.env, XDG_RUNTIME_DIR: root } });
  } finally {
    await fs.rm(root, { recursive: true, force: true });
  }
}

/** A file at a socket's expected path that answers no `net.connect` at all —
 * enough for `listHookSockets` to see it, not enough for anything to answer. */
async function plantDeadSocket(opts) {
  const loc = socketLocation(opts);
  await fs.mkdir(loc.dir, { recursive: true, mode: 0o700 });
  const p = loc.pathFor(loc.nameFor(999999));
  await fs.writeFile(p, '');
  return p;
}

// -------------------------------------------------------------- formatStatusLine

test('formatStatusLine renders incident + counts, omitting zero-valued segments', () => {
  assert.equal(
    formatStatusLine({ incidentId: 'inc-1', pending: 3, votesAwaited: 1 }),
    '🔴 landfall #inc-1 · 3 new · 1 vote awaited',
  );
  assert.equal(formatStatusLine({ incidentId: 'inc-1', pending: 0, votesAwaited: 0 }), '🔴 landfall #inc-1');
  assert.equal(
    formatStatusLine({ incidentId: 'inc-1', pending: 1, votesAwaited: 2 }),
    '🔴 landfall #inc-1 · 1 new · 2 votes awaited',
  );
});

test('formatStatusLine renders the divergence segment when present (T050 dependency)', () => {
  const line = formatStatusLine({
    incidentId: 'inc-1',
    pending: 0,
    votesAwaited: 0,
    divergence: { diverging: true, establishedSubject: 'cli-handoff/redeem', observedSubject: 'cloudfront/5xxerrorrate' },
  });
  assert.match(line, /diverging from established root cause/);
  assert.match(line, /cli-handoff\/redeem/);
});

test('formatStatusLine returns the empty string for null — the caller must print nothing', () => {
  assert.equal(formatStatusLine(null), '');
});

// -------------------------------------------------------------- queryStatus

test('queryStatus answers null when landfall serve is not running (no socket directory at all)', async () => {
  await withWorkspace(async (opts) => {
    assert.equal(await queryStatus(opts), null);
  });
});

test('queryStatus never reads a stale cache when nothing is running — a closed session must not bleed a stale incident into the statusline', async () => {
  await withWorkspace(async (opts) => {
    // A cache exists from an earlier, now-ended session...
    await writeCache({ ok: true, incidentId: 'inc-old', pending: 5, votesAwaited: 0 }, opts);
    // ...but there is no socket directory at all — serve is not running now.
    assert.equal(await queryStatus(opts), null);
  });
});

test('queryStatus falls back to the last-known cache when a socket exists but nothing answers (SC-012)', async () => {
  await withWorkspace(async (opts) => {
    await writeCache({ ok: true, incidentId: 'inc-1', pending: 2, votesAwaited: 1 }, opts);
    await plantDeadSocket(opts);
    const status = await queryStatus(opts);
    assert.equal(status.incidentId, 'inc-1');
    assert.equal(status.pending, 2);
  });
});

test('queryStatus answers null (not a throw) when a socket exists, nothing answers, and there is no cache either', async () => {
  await withWorkspace(async (opts) => {
    await plantDeadSocket(opts);
    assert.equal(await queryStatus(opts), null);
  });
});

test('queryStatus answers a real socket and caches what it got, end to end', async () => {
  await withWorkspace(async (opts) => {
    const session = {
      pending: [{ seq: 1 }, { seq: 2 }],
      cursor: 0,
      pendingDropped: 0,
      client: { cfg: { slug: 'acme', incidentId: 'inc-real' } },
      attention: { votesAwaited: [{ claimSeq: 1 }] },
    };
    // A short pid, deliberately: a real macOS `sun_path` limit (~104 bytes)
    // silently truncates a long socket filename's `.sock` suffix once the temp
    // path + workspace-key hash + a realistic multi-digit pid add up — found
    // live while writing this test (a 6-digit pid reproduced it every time,
    // `readdir` showed a bare `424242` with no extension on disk even though
    // `startHookSocket` reported success). That is a latent, pre-existing
    // fragility in socket.mjs's naming scheme, plausibly explaining some of
    // this repo's own pre-existing real-socket test failures — out of scope
    // for T048/T051 to fix, so this test just avoids tripping it.
    const handle = await startHookSocket(session, { ...opts, pid: 1 });
    assert.ok(handle, 'socket bind is expected to succeed in this environment');
    try {
      const status = await queryStatus(opts);
      assert.equal(status.incidentId, 'inc-real');
      assert.equal(status.pending, 2);
      assert.equal(status.votesAwaited, 1);
      // The real answer was cached, not just returned.
      assert.equal((await readCache(opts)).incidentId, 'inc-real');
    } finally {
      await handle.close();
    }
  });
});

test('readCache/writeCache round-trip and are unreadable-safe', async () => {
  await withWorkspace(async (opts) => {
    assert.equal(await readCache(opts), null); // nothing written yet
    await writeCache({ ok: true, incidentId: 'inc-9' }, opts);
    assert.equal((await readCache(opts)).incidentId, 'inc-9');
  });
});

// -------------------------------------------------------------- runStatus (the CLI command itself)

test('runStatus writes nothing and does not throw when landfall serve is not running — T051', async () => {
  await withWorkspace(async (opts) => {
    const written = [];
    const orig = process.stdout.write;
    process.stdout.write = (chunk) => { written.push(chunk); return true; };
    try {
      await runStatus(opts);
    } finally {
      process.stdout.write = orig;
    }
    assert.deepEqual(written, []);
  });
});

test('runStatus writes exactly one line when the socket answers', async () => {
  await withWorkspace(async (opts) => {
    await writeCache({ ok: true, incidentId: 'inc-1', pending: 0, votesAwaited: 0 }, opts);
    await plantDeadSocket(opts); // exists but unreachable — exercises the fallback path end to end
    const written = [];
    const orig = process.stdout.write;
    process.stdout.write = (chunk) => { written.push(chunk); return true; };
    try {
      await runStatus(opts);
    } finally {
      process.stdout.write = orig;
    }
    assert.equal(written.length, 1);
    assert.match(written[0], /inc-1/);
  });
});

File: internal/cli/status_test.go
// status_test.go — a port of `test/status.test.mjs` (11 cases), feature
// 20260812-010632 (US5/T048, T051).
//
// A genuine StartHookSocket+SendToSocket round trip works standalone here, so
// the success path binds a real socket. The fallback/absent-session paths
// deliberately do NOT need one: "no socket at all" needs only a directory read,
// and "a socket exists but nothing answers" is reproduced with a plain file at
// the expected socket path, which fails to dial fast without anything needing to
// be listening — smaller and more deterministic than racing a real server's
// shutdown timing.
package cli

import (
	"bytes"
	"context"
	"io"
	"os"
	"strings"
	"testing"

	"github.com/landfalls-ai/landfall-cli/internal/client"
	"github.com/landfalls-ai/landfall-cli/internal/hooks"
	"github.com/landfalls-ai/landfall-cli/internal/session"
)

// withWorkspace builds an isolated workspace under /tmp — short enough that a
// real socket path stays inside the 104-byte sockaddr_un.sun_path limit, which
// the default macOS $TMPDIR does not.
func withWorkspace(t *testing.T) hooks.Workspace {
	t.Helper()
	root, err := os.MkdirTemp("/tmp", "lf-status-")
	if err != nil {
		t.Fatalf("mkdtemp: %v", err)
	}
	t.Cleanup(func() { _ = os.RemoveAll(root) })
	return hooks.Workspace{Cwd: root, Env: map[string]string{"XDG_RUNTIME_DIR": root}}
}

// plantDeadSocket puts a file at a socket's expected path that answers no dial
// at all — enough for ListHookSockets to see it, not enough for anything to
// answer.
func plantDeadSocket(t *testing.T, ws hooks.Workspace) string {
	t.Helper()
	loc := hooks.SocketLocationFor(ws)
	if err := os.MkdirAll(loc.Dir, 0o700); err != nil {
		t.Fatal(err)
	}
	p := loc.PathFor(loc.NameFor("999999"))
	if err := os.WriteFile(p, nil, 0o600); err != nil {
		t.Fatal(err)
	}
	return p
}

func status(incidentID string, pending, votes int) hooks.SocketResponse {
	return hooks.SocketResponse{OK: true, IncidentID: incidentID, Pending: pending, VotesAwaited: votes}
}

// -------------------------------------------------------------- FormatStatusLine

func TestFormatStatusLineRendersIncidentAndCountsOmittingZeroSegments(t *testing.T) {
	s := status("inc-1", 3, 1)
	if got := FormatStatusLine(&s); got != "🔴 landfall #inc-1 · 3 new · 1 vote awaited" {
		t.Fatalf("got %q", got)
	}
	z := status("inc-1", 0, 0)
	if got := FormatStatusLine(&z); got != "🔴 landfall #inc-1" {
		t.Fatalf("got %q", got)
	}
	p := status("inc-1", 1, 2)
	if got := FormatStatusLine(&p); got != "🔴 landfall #inc-1 · 1 new · 2 votes awaited" {
		t.Fatalf("got %q", got)
	}
}

func TestFormatStatusLineRendersTheDivergenceSegmentWhenPresent(t *testing.T) {
	s := status("inc-1", 0, 0)
	s.Divergence = &client.Divergence{
		Diverging:          true,
		EstablishedSubject: "cli-handoff/redeem",
		ObservedSubject:    "cloudfront/5xxerrorrate",
	}
	line := FormatStatusLine(&s)
	if !strings.Contains(line, "diverging from established root cause") || !strings.Contains(line, "cli-handoff/redeem") {
		t.Fatalf("got %q", line)
	}
}

func TestFormatStatusLineIgnoresANonDivergingAnswer(t *testing.T) {
	s := status("inc-1", 0, 0)
	s.Divergence = &client.Divergence{Diverging: false}
	if got := FormatStatusLine(&s); got != "🔴 landfall #inc-1" {
		t.Fatalf("got %q", got)
	}
}

func TestFormatStatusLineReturnsEmptyForNilSoTheCallerPrintsNothing(t *testing.T) {
	if got := FormatStatusLine(nil); got != "" {
		t.Fatalf("got %q", got)
	}
}

// -------------------------------------------------------------- QueryStatus

func TestQueryStatusAnswersNilWhenServeIsNotRunning(t *testing.T) {
	if got := QueryStatus(withWorkspace(t)); got != nil {
		t.Fatalf("got %+v, want nil", got)
	}
}

func TestQueryStatusNeverReadsAStaleCacheWhenNothingIsRunning(t *testing.T) {
	ws := withWorkspace(t)
	// A cache exists from an earlier, now-ended session...
	WriteStatusCache(status("inc-old", 5, 0), ws)
	// ...but there is no socket directory at all — serve is not running now, and
	// a closed terminal's incident must not bleed into a blank statusline.
	if got := QueryStatus(ws); got != nil {
		t.Fatalf("got %+v, want nil", got)
	}
}

func TestQueryStatusFallsBackToTheCacheWhenASocketExistsButNothingAnswers(t *testing.T) {
	ws := withWorkspace(t)
	WriteStatusCache(status("inc-1", 2, 1), ws)
	plantDeadSocket(t, ws)

	got := QueryStatus(ws)
	if got == nil || got.IncidentID != "inc-1" || got.Pending != 2 {
		t.Fatalf("got %+v", got)
	}
}

func TestQueryStatusAnswersNilWhenNothingAnswersAndThereIsNoCacheEither(t *testing.T) {
	ws := withWorkspace(t)
	plantDeadSocket(t, ws)
	if got := QueryStatus(ws); got != nil {
		t.Fatalf("got %+v, want nil", got)
	}
}

func TestQueryStatusAnswersARealSocketAndCachesWhatItGot(t *testing.T) {
	ws := withWorkspace(t)
	s := session.New(session.Options{
		Client: client.New(client.Config{IncidentID: "inc-real", Slug: "acme"}, nil),
	})
	s.SetAttention(&client.Attention{VotesAwaited: []client.VoteAwaited{{}}})
	s.SetAttentionDirty(false)
	one, two := int64(1), int64(2)
	s.EnqueueEvent(context.Background(), client.Event{Seq: &one, Type: "edge.finding"})
	s.EnqueueEvent(context.Background(), client.Event{Seq: &two, Type: "edge.finding"})

	// A short pid, deliberately: the sockaddr_un.sun_path limit makes a long
	// socket filename the difference between a bound socket and a silent
	// failure.
	bound := hooks.StartHookSocket(context.Background(), s, hooks.StartOptions{Workspace: ws, PID: 1})
	if bound == nil {
		t.Fatal("socket bind is expected to succeed in this environment")
	}
	defer func() { _ = bound.Close() }()

	got := QueryStatus(ws)
	if got == nil || got.IncidentID != "inc-real" || got.Pending != 2 || got.VotesAwaited != 1 {
		t.Fatalf("got %+v", got)
	}
	// The real answer was CACHED, not just returned.
	cached := ReadStatusCache(ws)
	if cached == nil || cached.IncidentID != "inc-real" {
		t.Fatalf("cached %+v", cached)
	}
}

func TestStatusCacheRoundTripsAndIsUnreadableSafe(t *testing.T) {
	ws := withWorkspace(t)
	if got := ReadStatusCache(ws); got != nil {
		t.Fatalf("nothing written yet, got %+v", got)
	}
	WriteStatusCache(status("inc-9", 0, 0), ws)
	got := ReadStatusCache(ws)
	if got == nil || got.IncidentID != "inc-9" {
		t.Fatalf("got %+v", got)
	}
}

func TestReadStatusCacheAnswersNilForGarbageRatherThanFailing(t *testing.T) {
	ws := withWorkspace(t)
	WriteStatusCache(status("inc-9", 0, 0), ws)
	if err := os.WriteFile(statusCachePath(ws), []byte("{not json"), 0o600); err != nil {
		t.Fatal(err)
	}
	if got := ReadStatusCache(ws); got != nil {
		t.Fatalf("got %+v, want nil", got)
	}
}

// -------------------------------------------------- RunStatus (the command itself)

func TestRunStatusWritesNothingAndSucceedsWhenServeIsNotRunning(t *testing.T) {
	var out bytes.Buffer
	ui := &UI{Out: &out, Err: io.Discard}
	if code := RunStatus(ui, withWorkspace(t)); code != 0 {
		t.Fatalf("exit code %d, want 0 — a statusline must never surface an error", code)
	}
	if out.Len() != 0 {
		t.Fatalf("wrote %q, want silence", out.String())
	}
}

func TestRunStatusWritesExactlyOneLineWhenThereIsSomethingToSay(t *testing.T) {
	ws := withWorkspace(t)
	WriteStatusCache(status("inc-1", 0, 0), ws)
	plantDeadSocket(t, ws) // exists but unreachable — the fallback path, end to end

	var out bytes.Buffer
	ui := &UI{Out: &out, Err: io.Discard}
	if code := RunStatus(ui, ws); code != 0 {
		t.Fatalf("exit code %d, want 0", code)
	}
	if !strings.Contains(out.String(), "inc-1") {
		t.Fatalf("got %q", out.String())
	}
	if strings.Contains(out.String(), "\n") {
		t.Fatalf("a statusline command must write one line with no trailing newline, got %q", out.String())
	}
}
